### PR TITLE
fidelityAPI.py Holdings Update.

### PR DESCRIPTION
**Pull request by ImNotOssy**
This should allow fidelity to be able to grab complete holding data. 

REQUIRES YOU TO UNCHECK THIS:

image

### CODE REVIEW

**Comment by matthew55** [comment]
Thank you for the PR. If a setting is required to be checked, I would like it to be automated by the script so there is less confusion for the nontechnical user. You can add this, or if you like, this is something I would be willing to work on.

**Comment by ImNotOssy** [comment]
> Thank you for the PR. If a setting is required to be checked, I would like it to be automated by the script so there is less confusion for the nontechnical user. You can add this, or if you like, this is something I would be willing to work on.

i'll see what i can do, fidelity already slow enough doing the 2fa check, another check would be its downfall.

**Comment by matthew55** [comment]
> > Thank you for the PR. If a setting is required to be checked, I would like it to be automated by the script so there is less confusion for the nontechnical user. You can add this, or if you like, this is something I would be willing to work on.
> 
> i'll see what i can do, fidelity already slow enough doing the 2fa check, another check would be its downfall.

I hear what you are saying and I even did a double take for a second when I realized we might have to be navigating several complex menus of Fidelity's setting menus to implement this. However, luckily, we we are already on the page where the setting is and it doesn't even really have to lead so I am willing to add the extra 1-2 seconds. But yes, slow webpages suck for a tool like this.

**Comment by ImNotOssy** [comment]
> > > Thank you for the PR. If a setting is required to be checked, I would like it to be automated by the script so there is less confusion for the nontechnical user. You can add this, or if you like, this is something I would be willing to work on.
> > 
> > 
> > i'll see what i can do, fidelity already slow enough doing the 2fa check, another check would be its downfall.
> 
> I hear what you are saying and I even did a double take for a second when I realized we might have to be navigating several complex menus of Fidelity's setting menus to implement this. However, luckily, we we are already on the page where the setting is and it doesn't even really have to lead so I am willing to add the extra 1-2 seconds. But yes, slow webpages suck for a tool like this.

You log in, press positions, then the settings will be on the right, and click uncheck, then apply. tbh is not bad and settings stick, its account based, not cookies or data. I'll add a check.

**Comment by NelsonDane** [comment]
I agree that it should be automated, one less thing for people to miss and we have to troubleshoot :)

**Comment by ImNotOssy** [comment]
> I agree that it should be automated, one less thing for people to miss and we have to troubleshoot :)

I know exactly who you are talking about.

**Comment by ImNotOssy** [comment]
Seems like a complicated thing to try and automate, its all javascript (which im not familar with) and not one element change is visible if the box is unchecked or not.

**Comment by matthew55** [comment]
No worries, I will take a look at it and see what I can make happen. Maybe we will have to go without it after all

**Comment by matthew55** [comment]
@NelsonDane Can you test this? I don't have any holdings in my Fidelity account. Make sure you *don't* have the setting unchecked so that you can test the functionality I just added too.

**Comment by NelsonDane** [comment]
A few notes:

@matthew55 Great job with the auto-settings click, that worked great for me!

However, it failed for me because not everyone has their columns in the same order. Fidelity actually lets you click and drag columns positions around, so hardcoding to search at columns 2 + 10 for price and quantity respectively fails for me because I moved mine to be closer:
image
So for me, it would have to be 2 and 3 for quantity/price. I haven't looked at this PR in depth to know if that's a hard change, but hopefully not.

Then lastly once it's working I'd remove all the extra print logging so that it's silent unless error like normal. Thanks so much! @matthew55 @ImNotOssy

**Comment by ImNotOssy** [comment]
> A few notes:
> 
> @matthew55 Great job with the auto-settings click, that worked great for me!
> 
> However, it failed for me because not everyone has their columns in the same order. Fidelity actually lets you click and drag columns positions around, so hardcoding to search at columns 2 + 10 for price and quantity respectively fails for me because I moved mine to be closer:
> image
> So for me, it would have to be 2 and 3 for quantity/price. I haven't looked at this PR in depth to know if that's a hard change, but hopefully not.
> 
> Then lastly once it's working I'd remove all the extra print logging so that it's silent unless error like normal. Thanks so much! @matthew55 @ImNotOssy 

You know I love my print statements for debugging.

**Review by braydio — APPROVED** [review]

File: fidelityAPI.py [hunk 1]
```diff
@@ -281,38 +294,58 @@ def fidelity_holdings(fidelity_o: Brokerage, loop=None):
...
+                        stock_ticker = row.find_element(By.XPATH, ".//div[1]//button").text
+
+                        # Skip rows without a stock ticker (e.g., cash rows)
+                        if not stock_ticker:
+                            print("Empty stock ticker found, skipping this row.")
+                            continue
+
+                        # Retrieve and clean the data
+                        price_raw = row.find_element(By.XPATH, ".//div[2]/div/span").text
+                        quantity_raw = row.find_element(By.XPATH, ".//div[10]/div/span").text
```
Comment (NelsonDane) [inline fidelityAPI.py:319]: Make these dynamic instead of hardcoded values

**Review by NelsonDane — CHANGES_REQUESTED** [review]

**Comment by NelsonDane** [comment]
This works great, thanks @ImNotOssy. I cleaned it up, can you or @matthew55 confirm this works, then I'll merge. Thanks!

**Comment by matthew55** [comment]
I do not contain any holdings in Fidelity, so I can not test fully, but I was not getting any errors for normal or headless.

**Comment by braydio** [comment]
I'll give this a shot, I've got 1 account of each type. Will provide update in a moment

**Comment by braydio** [comment]
Looks like selenium can't find xpath it's looking for, I'm guessing one of the position detail rows? I've gotta run to an appointment but I can try and take a closer look when I get back. Attaching the (I believe) relevant console log. 
[PR#335 Console Log.md](https://github.com/user-attachments/files/16787670/PR.335.Console.Log.md)

**Comment by matthew55** [comment]
> Looks like selenium can't find xpath it's looking for, I'm guessing one of the position detail rows? I've gotta run to an appointment but I can try and take a closer look when I get back. Attaching the (I believe) relevant console log. [PR#335 Console Log.md](https://github.com/user-attachments/files/16787670/PR.335.Console.Log.md)

It looks like the change from:

```python
price_raw = row.find_element(By.XPATH, ".//div[2]/div/span").text
```
to:
```python
price_raw = row.find_element(By.XPATH, ".//div[@col-id='lstPrc']//span[@class='ag-cell-value']").text
```

Is causing the `Unexpected error processing row, skipping it: Message: no such element: Unable to locate element: {"method":"xpath","selector":".//div[@col-id='lstPrc']//span[@class='ag-cell-value']"}` as seen in your log file.

There must be something up with the new dynamic xpath.

**Comment by ImNotOssy** [comment]
> > Looks like selenium can't find xpath it's looking for, I'm guessing one of the position detail rows? I've gotta run to an appointment but I can try and take a closer look when I get back. Attaching the (I believe) relevant console log. [PR#335 Console Log.md](https://github.com/user-attachments/files/16787670/PR.335.Console.Log.md)
> 
> It looks like the change from:
> 
> ```python
> price_raw = row.find_element(By.XPATH, ".//div[2]/div/span").text
> ```
> to:
> ```python
> price_raw = row.find_element(By.XPATH, ".//div[@col-id='lstPrc']//span[@class='ag-cell-value']").text
> ```
> 
> Is causing the `Unexpected error processing row, skipping it: Message: no such element: Unable to locate element: {"method":"xpath","selector":".//div[@col-id='lstPrc']//span[@class='ag-cell-value']"}` as seen in your log file.
> 
> There must be something up with the new dynamic xpath.

😭😭

**Comment by ImNotOssy** [comment]
works fine, i was about to blame nelsons changes. but it seems to work fine for me.

**Comment by ImNotOssy** [comment]
@braydio do you mind sending a screenshot of how your columns look like on fidelity. like this
image

**Comment by matthew55** [comment]
If that is the case, @braydio can you please provide us more information. How are you running this (console, docker, discord)? Do you have headless mode on? How many holdings do you have in your accounts. How many accounts do you have on Fidelity? Thank you.

**Comment by ImNotOssy** [comment]
> If that is the case, @braydio can you please provide us more information. How are you running this (console, docker, discord)? Do you have headless mode on? How many holdings do you have in your accounts. How many accounts do you have on Fidelity? Thank you.

I think i saw headless was set to true.

**Comment by ImNotOssy** [comment]
image

**Comment by braydio** [comment]
Sure I've got 4 accounts on Fidelity that I do r/s in and 2 accounts that are 401k from my employer that do not have penny stock trading enabled. To run the script I checked out the PR in vscode and started it with python in the vscode terminal with the local venv. That starts up the discord bot where I entered the !rsa holdigns fidelity. It's running headless, and I've got 10 holdings max in the r/s accounts. It does look to be pulling this error 6 times, which would be one for each account (regardless of ability to trade penny stocks). Attached screenshot is the order I have my columns in. 

image

**Comment by braydio** [comment]
Well this is odd, searching the xpath in dev mode on my browser pulls the correct element. I'll check this out again and see if anything weird is going on when it runs

**Comment by ImNotOssy** [comment]
> Well this is odd, searching the xpath in dev mode on my browser pulls the correct element. I'll check this out again and see if anything weird is going on when it runs

most def a headless issue :/

**Comment by braydio** [comment]
Just an idea but, if you navigate to All Accounts > Positions and click the download link, it downloads a .CSV that appears to have a static column order with Current Value in the 9th column. It doesn't seem to change the column order in the csv download if you rearrange your column order on the website.

**Comment by NelsonDane** [comment]
> > Well this is odd, searching the xpath in dev mode on my browser pulls the correct element. I'll check this out again and see if anything weird is going on when it runs
> 
> most def a headless issue :/

If it was a headless issue, then it would work when `HEADLESS=False` and not when `HEADLESS=True`. Is this true for you?

**Comment by NelsonDane** [comment]
> Just an idea but, if you navigate to All Accounts > Positions and click the download link, it downloads a .CSV that appears to have a static column order with Current Value in the 9th column. It doesn't seem to change the column order in the csv download if you rearrange your column order on the website.

It's a good idea, but I'd rather not download a CSV every time the holdings command is ran. We'd have to add functions then to parse/verify the CSV data which would be a lot of extra work.

**Comment by braydio** [comment]
@matthew55 @ImNotOssy is the top account in your summary view a margin account or a cash account? Mine is margin and I had a thought that it might be causing an issue

**Comment by NelsonDane** [comment]
What happened @ImNotOssy

**Comment by ImNotOssy** [comment]
oh, i didnt mean to do that :0

**Pull request closed without merging**